nodejs/eks/nodegroup.ts

```diff
@@ -454,31 +454,44 @@ ${customUserData}
         minInstancesInService = 0;
     }
 
-    // Set default Name and Kubernetes tags on the ASG's.
-    let autoScalingGroupTags = pulumi.all([
-        eksCluster.name,
-    ]).apply(([clusterName]) => `- Key: Name
-                            Value: ${clusterName}-worker
-                            PropagateAtLaunch: 'true'
-                          - Key: kubernetes.io/cluster/${clusterName}
-                            Value: 'owned'
-                            PropagateAtLaunch: 'true'`);
-
-    // Merge any cluster tags for the ASG's.
-    if (core.tags) {
-        autoScalingGroupTags = pulumi.concat(autoScalingGroupTags, tagsToAsgTags(core.tags));
-    }
-    // Merge any user supplied tags for the ASG's.
+    // Assemble the ASG tags in the proper order to dedupe tags that are inherited,
+    // by using the right-most (last) k/v (in the chance of key collisions)
+    // during the merge of the InputTags maps (via the spread/splat operator):
+    // - core.tags
+    // - args.cloudFormationTags (AWS CF stack auto-tags the ASG w/ all of the CF tags)
+    // - args.autoScalingGroupTags
+    // - Name & clusterName required for Kubernetes
+    //
+    // Since AWS auto-tags the ASG with the tags it inherits from the
+    // CloudFormation stack, we must first dedupe args.autoScalingGroupTags of
+    // any conflicting tags that the CF stack already has / will set on the ASG.
+    let dedupedAsgTags: InputTags = <InputTags>{};
     if (args.autoScalingGroupTags) {
-        autoScalingGroupTags = pulumi.concat(autoScalingGroupTags, tagsToAsgTags(args.autoScalingGroupTags));
+        dedupedAsgTags = args.autoScalingGroupTags;
+        if (args.cloudFormationTags) {
+            dedupedAsgTags = dedupeInputTags(args.cloudFormationTags, args.autoScalingGroupTags);
+        }
     }
+    const autoScalingGroupTags: InputTags = pulumi.all([
+        eksCluster.name,
+        core.tags,
+        dedupedAsgTags,
+    ]).apply(([clusterName, tags, asgTags]) => ({
+        ...tags,
+        ...asgTags,
+        "Name": `${clusterName}-worker`,
+        [`kubernetes.io/cluster/${clusterName}`]: "owned",
```

Comment:
Now that we are merging these tags with `{...}`, which will take care of overriding keys, why do we need any of the `dedupe` logic above?  Won't that be taken care of automatically by this object merge?

Reply:
Also - is this the right order for the merges?  I assume we want to let users override these defaults if really needed - especially things like `Name`?  Any reason not to apply `...tags` and `...asgTags` last?

Reply:
> Won't that be taken care of automatically by this object merge?

@lukehoban You are correct that merging maps with the spread will take care of overriding keys. However, imagine this example scenario:
- We set `cloudFormationTags` to be: `{ env: "dev", myCloudFormationTag: "true" }`
- We set `autoScalingGroupTags` to be:  `{ env: "dev", myAutoScalingGroupTag: "true" }`

If we included the `cloudFormationTags` in the map merge + spread deduping being done above for `autoScalingGroupTags`, the `autoScalingGroupTags` to set now become:
 `{ myAutoScalingGroupTag: "true", myCloudFormationTag: "true" }` since `env: "dev"` is a dupe and omitted, and `myCloudFormationTag: "true"` is not a dupe, so it's included.

This means that when the ASG tries to set `myCloudFormationTag: "true"` as one of its tags, it will conflict with AWS, as AWS will try to set this tag on both the CF Stack *and* the ASG automatically - so the CF + ASG will hit a dupe tag scenario.

This is why we use the the logic `dedupeInputTags(args.cloudFormationTags, args.autoScalingGroupTags);` as it will only return tags in `args.autoScalingGroupTags` that do not exist in `args.cloudFormationTags` - as AWS will auto include the CF tags on the ASG for us.

Do you agree?